### PR TITLE
dai: Remove frame_fmt from private data

### PR DESCRIPTION
**Pull request by ktrzcinx**
This value is one of audio_stream property, so should be
saved in this place to keep data synchronised, and allow
to usage generic function working on audio_stream, like
stream converters.

Signed-off-by: Karol Trzcinski <karolx.trzcinski@linux.intel.com>

### CODE REVIEW

**Comment by ktrzcinx** [comment]
SOFCI TEST

File: src/audio/dai.c [hunk 1]
```diff
@@ -473,6 +473,14 @@ static int dai_params(struct comp_dev *dev,
 			comp_err(dev, "dai_params(): failed to alloc dma buffer");
 			return -ENOMEM;
 		}
+
+		/*
```
Comment (mmaka1) [inline src/audio/dai.c:480]: The comment applies to the next line, could be rephrased a bit to refer to hw-params (and req'ed dma buffer settings) vs. params.

File: src/audio/dai.c [hunk 2]
```diff
@@ -473,6 +473,14 @@ static int dai_params(struct comp_dev *dev,
...
+		 * dma_buffer frame_fmt should be overwritten with value
+		 * received in sof_ipc_comp_config, because value from params
+		 * represents data format for component connected with dai
+		 */
+		buffer_set_params(dd->dma_buffer, params, true);
+		dd->dma_buffer->stream.frame_fmt = dconfig->frame_fmt;
```
Comment (mmaka1) [inline src/audio/dai.c:486]: This is unconditional override while dai_comp_get_hw_params() tests != 0. Is this asymmetry intentional? If so, please also explain inside the comment.

Comment (mmaka1) [inline src/audio/dai.c:486]: Simple oneliner audio_stream_override_frame_fmt() could be helpful (1/ easier to grep, 2/ improves encapsulation).

File: src/audio/dai.c [hunk 3]
```diff
@@ -473,6 +473,14 @@ static int dai_params(struct comp_dev *dev,
 			comp_err(dev, "dai_params(): failed to alloc dma buffer");
 			return -ENOMEM;
 		}
+
+		/*
+		 * dma_buffer frame_fmt should be overwritten with value
+		 * received in sof_ipc_comp_config, because value from params
+		 * represents data format for component connected with dai
+		 */
+		buffer_set_params(dd->dma_buffer, params, true);
```
Comment (mmaka1) [inline src/audio/dai.c:485]: There is dd->dma_buffer->buffer_fmt update. Pls remove if this one is unused (vs. inner audio stream frame fmt).

File: src/audio/dai.c (hunk 1 above)
Comment (ktrzcinx) [reply src/audio/dai.c:479]: rephrased

File: src/audio/dai.c (hunk 2 above)
Comment (ktrzcinx) [reply src/audio/dai.c]: removed condition from dai_comp_get_hw_params()

Comment (ktrzcinx) [reply src/audio/dai.c]: added

File: src/audio/dai.c (hunk 3 above)
Comment (ktrzcinx) [reply src/audio/dai.c:485]: https://github.com/thesofproject/sof/pull/3179

File: src/audio/dai.c (hunk 2 above)
Comment (mmaka1) [reply src/audio/dai.c]: Another solution would be to create a local copy of params, override frame_fmt and just set with a single audio-stream API call.

**Comment by lgirdwood** [comment]
What's the rational for all the buffer/stream changes today ?

**Comment by ktrzcinx** [comment]
>What's the rational for all the buffer/stream changes today ?

After https://github.com/thesofproject/sof/issues/3146 I took a closer look to dma_buffers usage, and I saw that there are sometimes treated as usual `audio_stream` buffers (`pcm_converter`, invalidation, writebacks) and sometimes like special, dma buffer, without any special data type (unset stream parameters like fmt, rate, channels). This is quite error prone. Moreover `audio_stream` is newer than `host`/`dai` implementation, so there are places where audio_stream API may be used, instead doing the same thing manually.

**Comment by ktrzcinx** [comment]
>Another solution would be to create a local copy of params, override frame_fmt and just set with a single audio-stream API call.

Then new API function with single user won't be needed (`buffer_set_frame_fmt()`), updated to this version

**Comment by plbossart** [comment]
> > What's the rational for all the buffer/stream changes today ?
> 
> After #3146 I took a closer look to dma_buffers usage, and I saw that there are sometimes treated as usual `audio_stream` buffers (`pcm_converter`, invalidation, writebacks) and sometimes like special, dma buffer, without any special data type (unset stream parameters like fmt, rate, channels). This is quite error prone. Moreover `audio_stream` is newer than `host`/`dai` implementation, so there are places where audio_stream API may be used, instead doing the same thing manually.

it may be error prone, but by the same token we should not assume that we know how many channels a DMA buffer contains. It could contain binary data or compressed data unrelated to PCM, or compressed data padded to fit the PCM rate. 

DMAs are general purpose tools to move data from point A to point B, we don't want to assume we only transfer PCM.

**Comment by ktrzcinx** [comment]
I agree, from DMA point of view, it doesn't matter what is the format of a data, and here such an information should be useless. Looking from `pcm_converter` point of view, stream metadata should be set (it's general purpose tool, not DMA specific).
So we need to have a point where binary data from DMA becomes an `audio_stream`, it should be done before `pcm_converter` usage. So far, this point is hidden somewhere in `host` component, during compressed audio stream implementation, it will be defined explicit, but it's not know.

**Comment by paulstelian97** [comment]
> I agree, from DMA point of view, it doesn't matter what is the format of a data, and here such an information should be useless. Looking from `pcm_converter` point of view, stream metadata should be set (it's general purpose tool, not DMA specific).
> So we need to have a point where binary data from DMA becomes an `audio_stream`, it should be done before `pcm_converter` usage. So far, this point is hidden somewhere in `host` component, during compressed audio stream implementation, it will be defined explicit, but it's not know.

Can't the `host` component be adapted to be more flexible and to allow for raw data to exit it whenever adequate? As in being able to set as params the format to "raw", some reserved value for raw data. And the decoder component would take in raw and push out proper PCM.

**Comment by plbossart** [comment]
> I agree, from DMA point of view, it doesn't matter what is the format of a data, and here such an information should be useless. Looking from `pcm_converter` point of view, stream metadata should be set (it's general purpose tool, not DMA specific).
> So we need to have a point where binary data from DMA becomes an `audio_stream`, it should be done before `pcm_converter` usage. So far, this point is hidden somewhere in `host` component, during compressed audio stream implementation, it will be defined explicit, but it's not know.

It seems we are in agreement on the basics, but your comment about the host component threw me off. This PR changes the DAIs,

**Comment by paulstelian97** [comment]
> > I agree, from DMA point of view, it doesn't matter what is the format of a data, and here such an information should be useless. Looking from `pcm_converter` point of view, stream metadata should be set (it's general purpose tool, not DMA specific).
> > So we need to have a point where binary data from DMA becomes an `audio_stream`, it should be done before `pcm_converter` usage. So far, this point is hidden somewhere in `host` component, during compressed audio stream implementation, it will be defined explicit, but it's not know.
> 
> It seems we are in agreement on the basics, but your comment about the host component threw me off. This PR changes the DAIs,

I chimed in with the host component because the DAI itself should always have PCM data (well, almost always; some DAIs can support other formats but it's not typical). Never compressed data. Sure, the DAI must have its own configuration readily available for the params propagation.

**Comment by plbossart** [comment]
> > > I agree, from DMA point of view, it doesn't matter what is the format of a data, and here such an information should be useless. Looking from `pcm_converter` point of view, stream metadata should be set (it's general purpose tool, not DMA specific).
> > > So we need to have a point where binary data from DMA becomes an `audio_stream`, it should be done before `pcm_converter` usage. So far, this point is hidden somewhere in `host` component, during compressed audio stream implementation, it will be defined explicit, but it's not know.
> > 
> > 
> > It seems we are in agreement on the basics, but your comment about the host component threw me off. This PR changes the DAIs,
> 
> I chimed in with the host component because the DAI itself should always have PCM data (well, almost always; some DAIs can support other formats but it's not typical). Never compressed data. Sure, the DAI must have its own configuration readily available for the params propagation.

It's perfectly possible to send/receive data as pseudo-PCM but the contents is packetized and could be variable bitrate internally.
People have abused isochronous interfaces big time in the past, and will continue doing so...
There's also cases where audio was sent on SPI, e.g. for bursty data or to flush a history buffer.
Don't confuse transmission/link and contents, they can differ.

**Comment by lgirdwood** [comment]
CI clipping. @aiChaoSONG difficult to tell from sof-logger and topology output which PGA is > 0dB
https://sof-ci.01.org/sofpr/PR3174/build6644/devicetest/

```
[ 98762163.802083] (     1635.937500) c0 pga          3.20  ../audio/volume/volume.c:641  volume_ctrl_get_cmd(), channel = 0, value = 65536
[ 98762219.010417] (       55.208332) c0 pga          3.20  ../audio/volume/volume.c:641  volume_ctrl_get_cmd(), channel = 1, value = 65536
[ 98764363.072917] (     2144.062500) c0 pga          2.13  ../audio/volume/volume.c:641  volume_ctrl_get_cmd(), channel = 0, value = 65536
[ 98764418.281250] (       55.208332) c0 pga          2.13  ../audio/volume/volume.c:641  volume_ctrl_get_cmd(), channel = 1, value = 65536
[ 98766301.354167] (     1883.072876) c0 pga          1.3   ../audio/volume/volume.c:641  volume_ctrl_get_cmd(), channel = 0, value = 65536
[ 98766356.562500] (       55.208332) c0 pga          1.3   ../audio/volume/volume.c:641  volume_ctrl_get_cmd(), channel = 1, value = 65536
[ 98768245.885417] (     1889.322876) c0 pga          1.2   ../audio/volume/volume.c:641  volume_ctrl_get_cmd(), channel = 0, value = 65536
[ 98768301.093750] (       55.208332) c0 pga          1.2   ../audio/volume/volume.c:641  volume_ctrl_get_cmd(), channel = 1, value = 65536
[102194911.302083] (  3426610.250000) c0 pga          1.2   ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 0, value = 65536
[102194976.822917] (       65.520836) c0 pga          1.2   ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 1, value = 65536
[102258245.312500] (    63268.488281) c0 pga          1.3   ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 0, value = 65536
[102258310.833333] (       65.520836) c0 pga          1.3   ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 1, value = 65536
[102366153.697917] (   107842.867188) c0 pga          2.13  ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 0, value = 413542
[102366219.635417] (       65.937500) c0 pga          2.13  ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 1, value = 413542
[102370531.927083] (     4312.291504) c0 pga          3.20  ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 0, value = 65536
[102370597.447917] (       65.520836) c0 pga          3.20  ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 1, value = 65536
[148939389.635417] ( 46568792.000000) c0 pga          1.2   ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 0, value = 65536
[148939455.156250] (       65.520836) c0 pga          1.2   ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 1, value = 65536
[148985572.604167] (    46117.449219) c0 pga          1.3   ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 0, value = 65536
[148985638.125000] (       65.520836) c0 pga          1.3   ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 1, value = 65536
[149086056.406250] (   100418.281250) c0 pga          2.13  ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 0, value = 413542
[149086122.343750] (       65.937500) c0 pga          2.13  ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 1, value = 413542
[149184145.468750] (    98023.125000) c0 pga          3.20  ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 0, value = 65536
[149184210.989583] (       65.520836) c0 pga          3.20  ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 1, value = 65536
```
It looks like sof-logger is merging PGA 1.3 into 2.13 ???  Fwiw topology graph only shows PGA 1.0, 1.1, 2.0 and 3.0. So we have a +0dB gain still in some topology tests.

**Comment by lgirdwood** [comment]
@plbossart any comments with the update for compressed streams ?

**Comment by plbossart** [comment]
> @plbossart any comments with the update for compressed streams ?

There was no answer to my comments and I don't see any changes in the code?

**Comment by aiChaoSONG** [comment]
> CI clipping. @aiChaoSONG difficult to tell from sof-logger and topology output which PGA is > 0dB
> https://sof-ci.01.org/sofpr/PR3174/build6644/devicetest/
> 
> ```
> [ 98762163.802083] (     1635.937500) c0 pga          3.20  ../audio/volume/volume.c:641  volume_ctrl_get_cmd(), channel = 0, value = 65536
> [ 98762219.010417] (       55.208332) c0 pga          3.20  ../audio/volume/volume.c:641  volume_ctrl_get_cmd(), channel = 1, value = 65536
> [ 98764363.072917] (     2144.062500) c0 pga          2.13  ../audio/volume/volume.c:641  volume_ctrl_get_cmd(), channel = 0, value = 65536
> [ 98764418.281250] (       55.208332) c0 pga          2.13  ../audio/volume/volume.c:641  volume_ctrl_get_cmd(), channel = 1, value = 65536
> [ 98766301.354167] (     1883.072876) c0 pga          1.3   ../audio/volume/volume.c:641  volume_ctrl_get_cmd(), channel = 0, value = 65536
> [ 98766356.562500] (       55.208332) c0 pga          1.3   ../audio/volume/volume.c:641  volume_ctrl_get_cmd(), channel = 1, value = 65536
> [ 98768245.885417] (     1889.322876) c0 pga          1.2   ../audio/volume/volume.c:641  volume_ctrl_get_cmd(), channel = 0, value = 65536
> [ 98768301.093750] (       55.208332) c0 pga          1.2   ../audio/volume/volume.c:641  volume_ctrl_get_cmd(), channel = 1, value = 65536
> [102194911.302083] (  3426610.250000) c0 pga          1.2   ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 0, value = 65536
> [102194976.822917] (       65.520836) c0 pga          1.2   ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 1, value = 65536
> [102258245.312500] (    63268.488281) c0 pga          1.3   ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 0, value = 65536
> [102258310.833333] (       65.520836) c0 pga          1.3   ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 1, value = 65536
> [102366153.697917] (   107842.867188) c0 pga          2.13  ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 0, value = 413542
> [102366219.635417] (       65.937500) c0 pga          2.13  ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 1, value = 413542
> [102370531.927083] (     4312.291504) c0 pga          3.20  ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 0, value = 65536
> [102370597.447917] (       65.520836) c0 pga          3.20  ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 1, value = 65536
> [148939389.635417] ( 46568792.000000) c0 pga          1.2   ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 0, value = 65536
> [148939455.156250] (       65.520836) c0 pga          1.2   ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 1, value = 65536
> [148985572.604167] (    46117.449219) c0 pga          1.3   ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 0, value = 65536
> [148985638.125000] (       65.520836) c0 pga          1.3   ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 1, value = 65536
> [149086056.406250] (   100418.281250) c0 pga          2.13  ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 0, value = 413542
> [149086122.343750] (       65.937500) c0 pga          2.13  ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 1, value = 413542
> [149184145.468750] (    98023.125000) c0 pga          3.20  ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 0, value = 65536
> [149184210.989583] (       65.520836) c0 pga          3.20  ../audio/volume/volume.c:557  volume_ctrl_set_cmd(), channel = 1, value = 65536
> ```
> 
> It looks like sof-logger is merging PGA 1.3 into 2.13 ??? Fwiw topology graph only shows PGA 1.0, 1.1, 2.0 and 3.0. So we have a +0dB gain still in some topology tests.

@keqiaozhang could you please help to check?

**Comment by ktrzcinx** [comment]
@plbossart I'm not sure what you would like to know.
This PR moves stream metadata information (mainly frame_fmt) storage place from `DAI` component to `audio_stream` (not DMA), do you think it's bad direction ?
This `audio_stream` is used by DMA, but updated data fields are not used there, so from DMA point of view nothing has changed.
When compressed audio_stream will be passed through DMA, then DMA won't care whats inside, as it is right now (before and after my changes).

**Comment by plbossart** [comment]
> @plbossart I'm not sure what you would like to know.
> This PR moves stream metadata information (mainly frame_fmt) storage place from `DAI` component to `audio_stream` (not DMA), do you think it's bad direction ?
> This `audio_stream` is used by DMA, but updated data fields are not used there, so from DMA point of view nothing has changed.
> When compressed audio_stream will be passed through DMA, then DMA won't care whats inside, as it is right now (before and after my changes).

I was referring to this code: https://github.com/thesofproject/sof/pull/3174/commits/ba6a21ae9871f1fcca594aa65d964530f9775b96#diff-9f8de02ec82000817dd1a9bbccac158cR76

this still has DAIs handling samples, and it's rather odd to me.

**Comment by ktrzcinx** [comment]
As I understand you are talking about removing calculation of number of samples from `dai_dma_cb`. 
It wasn't added in this PR, but this one is one step forward to make it possible.
In target code it may be something like:
```
dma_buffer_copy_to(dd->local_buffer, dd->dma_buffer, bytes,
		   dd->decompress_data);
```
and
```
void dma_buffer_copy_to(struct comp_buffer *source, struct comp_buffer *sink, uint32_t source_bytes,
			struct converter_data *decompress)
{
	struct audio_stream *ostream = &sink->stream;
        int sink_bytes;

	buffer_invalidate(source, source_bytes);

	/* decompress data */
	sink_bytes = converter_process(&source->stream, 0, ostream, 0, source_bytes, decompress);
        if (sink_bytes < 0)
             // errro handling

	/* sink buffer contains data meant to copied to DMA */
	audio_stream_writeback(ostream, sink_bytes);

	ostream->w_ptr = (char *)ostream->w_ptr + sink_bytes;
	ostream->w_ptr = audio_stream_wrap(ostream, ostream->w_ptr);

	comp_update_buffer_consume(source, source_bytes);
}
```
where `pcm_converter` will be called internally in decompress_process (when needed). But as I said it before, this PR is not going to provide full compressed stream support, this one is only preparation for it.

**Comment by plbossart** [comment]
SOFCI TEST

**Comment by ktrzcinx** [comment]
Thanks @plbossart

**Pull request merged**